### PR TITLE
[receiver/prometheus] Don't fail the whole scrape on invalid data

### PR DESCRIPTION
**Pull request by swiatekm**
**Description:**
Prometheus receiver no longer fails the whole scrape when encountering invalid datapoints. Instead, it logs a warning and continues with the transaction.

**Link to tracking Issue:** #24030

**Testing:**
Modified existing transaction tests for validation errors and added a new one checking whether valid data in a transaction is committed even if invalid data was also appended.

I also had to make a change to the negative openmetrics tests. These tests assert that the scrape fails on invalid data, but in practice this happens for two different reasons, and this PR removes one of them. I split the failure case into the "fail" case, where the scrape fails completely, and the "invalid" case where the scrape succeeds, but invalid metrics are dropped.

### CODE REVIEW

File: receiver/prometheusreceiver/metrics_receiver_open_metrics_test.go [hunk 1]
```diff
@@ -84,7 +84,12 @@ func verifyNegativeTarget(t *testing.T, td *testData, mds []pmetric.ResourceMetr
 
 	require.Greater(t, len(mds), 0, "At least one resource metric should be present")
 	metrics := getMetrics(mds[0])
-	assertUp(t, 0, metrics)
+
+	// There are two possibilities here
+	// 1. The Prometheus scrape parser rejected the sample as invalid. In this case, we get no metrics.
+	// 2. The Prometheus scrape parser accepted the sample, but the receiver dropped it due to incompatibility with the Otel schema.
```
Comment (dashpole) [inline receiver/prometheusreceiver/metrics_receiver_open_metrics_test.go:90]: Should we just make case 2 a "positive" case?  Or should we separate into three verify function?

Comment (swiatekm-sumo) [reply receiver/prometheusreceiver/metrics_receiver_open_metrics_test.go:90]: Logically, it seems "negative" to me, in the sense that we don't actually ingest the metric. If anything the "positive" test should check that any non-default metric is present.

Comment (dashpole) [reply receiver/prometheusreceiver/metrics_receiver_open_metrics_test.go:90]: Could we split the negative cases into fail vs invalid then?

Comment (swiatekm-sumo) [reply receiver/prometheusreceiver/metrics_receiver_open_metrics_test.go]: Allright, I renamed the "invalid" test cases appropriately and added a separate test for them. The rest stays the same, although I've also added the additional check to positive tests - they all pass regardless.

**Review by dashpole — APPROVED** [review]

**Review by atoulme — APPROVED** [review]

**Comment by krisztianfekete** [comment]
Hey, I stumbled upon on this, and IIRC Prometheus doesn't have partial failures for resiliency purposes and because managing those is hard. I am wondering if this would diverge from the logic there, especially since this receiver is supposed to be a "drop-in" replacement for Prometheus.

**Comment by dashpole** [comment]
> Hey, I stumbled upon on this, and IIRC Prometheus doesn't have partial failures for resiliency purposes and because managing those is hard. I am wondering if this would diverge from the logic there, especially since this receiver is supposed to be a "drop-in" replacement for Prometheus.

It does diverge from that principle to some extent, but we have to choose between two incompatibilities.  We either always fail scraping endpoints that prometheus is able to scrape (current state), or we succeed in scraping those endpoints, but drop points we can't translate to our data model.  We do still completely fail on scrapes the prometheus server would fail on, but not on translation errors.  The current state makes it impossible to migrate some applications to the collector, which seems bad.  I suppose an alternative would be to make it configurable.

**Comment by swiatekm-sumo** [comment]
> Hey, I stumbled upon on this, and IIRC Prometheus doesn't have partial failures for resiliency purposes and because managing those is hard. I am wondering if this would diverge from the logic there, especially since this receiver is supposed to be a "drop-in" replacement for Prometheus.

Keep in mind, this change only affects errors from the Appender, we still use Prometheus' parsing infrastructure and reject everything it would reject for syntax failures.

At the very least, the receiver doesn't accept any data Prometheus rejects - this is the case both before and after this change. Prometheus doesn't really reject data for failure to conform to the spec, whereas the receiver does - this change basically allows those rejections to be partial.

**Comment by krisztianfekete** [comment]
Hey @swiatekm-sumo, @dashpole, thanks for the reply, I also checked the PR properly in the meantime, great!

**Comment by swiatekm-sumo** [comment]
Is there anything blocking merging this? It's been ready for more than a week.

**Review by codeboten — APPROVED** [review]

**Pull request merged**